### PR TITLE
Update ICU to 63.1, bump revision of all dependant ports

### PR DESCRIPTION
**Pull request by lbschenkel**
@ryandesign:
I did some basic testing on my machine, did not detect any breakage.
I bumped the revision of all dependant ports to force a rebuild. I cannot really test all of them.

### CODE REVIEW

**Comment by MarcusCalhoun-Lopez** [comment]
I do not believe this patch will work with the Qt ports.
Each sub-port has its own revision number, which will override any "global" revision.

**Comment by ryandesign** [comment]
The icu port has a revision declared in this block:

```
if {${subport} eq ${name}} {
    revision                2
}
```

This needs to be changed to `0`.

The boost port has a subport boost-numpy. Ports that have subports often set a different revision in each subport, as the boost port does. So you need to increase the two `revision` lines later on in the boost portfile, rather than adding a new `revision` line (that won't be used) earlier in the portfile.

In the php port, you only increased the revisions of the php53-intl, php54-intl, and php55-intl subports, because they already had a nonzero revision set. You need to increase the revision of every port using icu, including those that didn't have a revision set yet, in this case including php56-intl, php70-intl, php71-intl, and php72-intl.

**Comment by lbschenkel** [comment]
Thanks @ryandesign. Since I had to manually update a lot of ports, it became repetitive quite quickly and obviously I missed some things. Unfortunately I won't have much free time in the next couple of days, but I'm going to get back to this as soon as I can. If somebody doesn't want to wait and wants to finish this work in the meantime, feel free to do so. I don't mind at all.

**Comment by ci42** [comment]
@lbschenkel 
The nodejs8 and nodejs9 ports don't require a revbump. They don't use the ICU port.

**Comment by neverpanic** [comment]
+1 for openttd from me.

**Comment by pmetzger** [comment]
Tomorrow will be a week since this went in. Do we feel that will have been enough time for review? It will have been over twice the usual 72 hours.

**Comment by lbschenkel** [comment]
Unfortunately I have been busy this week and I was not able to work on this PR, but I'm hoping to pick it up again soon and finish the work. This PR is basically waiting for me, not the reviewers.

**Comment by pmetzger** [comment]
@lbschenkel Okay, I'm going to do two things:

1. I'm going to mark this WIP, indicating it should not be committed.
2. When you're ready to see this committed, please indicate it in the comments here.

Also, keep in mind that if you don't work on it for too long, you'll get nagging^wgentle reminders from me about it. 😃

**Comment by pmetzger** [comment]
@lbschenkel we now have conflicts, so if you could re-do this it would be appreciated.

**Comment by lbschenkel** [comment]
@ryandesign: fixed ICU port
@ci42: removed revbumps from `nodejs8` and `nodejs9`
@MarcusCalhoun-Lopez: fixed `qt5` port, however I'm unsure if I need to leave the revbump in the main port or not. I'm leaving the revbump for now, unless you tell me it's not necessary.

Sorry for the delay on getting back to this PR, I got unexpectedly busy. I also feel that I did a somewhat rushed job in the original submission, ended up being sub-par.

**Comment by lbschenkel** [comment]
@MarcusCalhoun-Lopez: done. I forgot to mention earlier that I deliberately just changed `qt5` and wanted you to take a look before I spent the time to change the other `qt*` ports.

**Comment by pmetzger** [comment]
@lbschenkel Given that you've removed WIP, does this mean you're ready to commit the change?

**Comment by lbschenkel** [comment]
@pmetzger: That is correct. From where I stand all outstanding issues were corrected. However, before merging I would prefer if @ryandesign gave the OK, since he's the maintainer.

**Comment by pmetzger** [comment]
@ryandesign are you happy with where this now is?

File: devel/icu/Portfile [hunk 1]
```diff
@@ -8,7 +8,7 @@ set my_name     icu4c
 # whenever the library version number changes. Thanks.
 # To find dependencies, use:
 # port file all | sort -u | xargs grep -El ':icu( |$)' | rev | cut -d / -f 2 | rev | sort -u
-version                     58.2
+version                     61.1
 # Don't update to 59 or later because they require C++11
```
Comment (ryandesign) [inline devel/icu/Portfile]: Please don't merge yet. As this comment points out, it does not build on systems that don't have C++11 support. Removing this comment line and adding the cxx11 1.1 portgroup lets icu build on 10.6, and I also successfully built harfbuzz-icu using it, but icu-lx failed to build. I haven't investigated further.

Comment (lbschenkel) [reply devel/icu/Portfile]: Thanks for pointing that out. I can't believe I missed this comment.

**Comment by lbschenkel** [comment]
I have rebased this branch and fixed the conflicts.

**Comment by michaelld** [comment]
If the buildbots say that Boost builds using the updated ICU, then I'm good with whatever changes are made to that port.

**Comment by pmetzger** [comment]
I'm unclear on where this is now. What has to be done next to get it to mergeable?

**Comment by pmetzger** [comment]
1. The branch now has conflicts, so those should be resolved.
2. This PR is starting to get quite old. What's needed to finish it?

**Comment by lbschenkel** [comment]
When I opened this PR, I was not aware of the C++11 issue and thought it was a more-or-less straightforward (but laborious) bump. @ryandesign pointed out that there is some breakage on other packages but I'm afraid that's beyond my current ability to diagnose/fix. I'm not sure if @ryandesign is or can pursue this at the moment. In case he intends to take this work and move forward I can solve the conflicts and push new changes, but as I said I'm not able to fix the side-effects of this bump on my own.

**Comment by ryandesign** [comment]
> What's needed to finish it?

Investigation of the consequences of icu using C++11 for other ports that use icu, most interestingly those that don't use C++11, on 10.8 and earlier, and correction of any issues found. I am not working on the problem.

**Comment by pmetzger** [comment]
Given that no one is working on the problem, I would propose that this be put in a branch, documented in a Trac ticket, and the pull request closed.

**Comment by kencu** [comment]
I'd commit it and then sort it out. Can't hold all MP back for ancient systems. I would not like to see more and more dead Macports branches sitting around.

**Comment by lbschenkel** [comment]
@ryandesign is the maintainer, I believe it should be your call. If you're OK with upgrading ICU and fixing eventual breakage later, I'll fix the conflicts and push so you can merge. I could contribute with that process if I'm told what to do (I don't have any ancient systems to do any testing/diagnosing/troubleshooting).

**Comment by michaelld** [comment]
I'm with Ken here: fix this PR's conflicts, commit it, then sort out issues with other ports & older OSs. Forcing C++11 isn't challenging these days.

**Comment by lbschenkel** [comment]
I want to add that due to the amount of ports that need to be compiled, Travis CI will always time out. I'm in the dark if I broke some of the ports I touched in this PR (I don't think so, but just saying I cannot be 100% sure). If/when this is merged we should keep an eye for regressions. Since I proposed the bump I'll do my best but I gladly appreciate any help.

**Comment by pmetzger** [comment]
@ryandesign are you okay with the proposed "commit and fix after" strategy? FYI, I also agree with @kencu and @michaelld on this.

File: devel/icu/Portfile [hunk 2]
```diff
@@ -1,15 +1,15 @@
 # -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:filetype=tcl:et:sw=4:ts=4:sts=4
 
 PortSystem      1.0
+PortGroup       cxx11 1.1
```
[the remark this answers is not in the document]
Comment (lbschenkel) [reply devel/icu/Portfile]: Fixed. Added only to the main and `lx` subport.

**Comment by ryandesign** [comment]
> The ICU LayoutEngine has been removed in ICU 58

That's right, the layout engine was removed from icu core in version 58 and moved to a separate icu-lx package. See a4c3703f5d6f350687dc372ec60b5b9b8cfd84d1.

**Comment by pmetzger** [comment]
We now have merge conflicts. Could those get resolved and could we decide to either merge this or close it?

**Comment by ryandesign** [comment]
I will work on it when I have time and interest in working on it.

**Comment by pmetzger** [comment]
> I will work on it when I have time and interest in working on it.

That makes it sound like this isn't really a pull request any more as it might happen in the far future.

**Comment by pmetzger** [comment]
@lbschenkel It looks like we're permanently stalled. How would you like to proceed? Should we just close this temporarily for now and re-open it some time in the future?

**Comment by lbschenkel** [comment]
It's a good question. I thought this was going to be a straightforward version bump. I would like to pursue this further, however I didn't have much free time lately to spend on this PR (the conflicts/bumps are not a problem, the issue is fixing `icu-lx`). I think I'll have more time to invest on this 2 weeks from now. I'll let you decide if you want to keep it open or not.

**Comment by pmetzger** [comment]
I suggest you re-open it in two weeks when you restart work, and push a resolution of the conflicts on the branch at that time.

**Comment by kencu** [comment]
@lbschenkel -- shall we get this going again?

**Comment by pmetzger** [comment]
@kencu If you wish to, you can just click "reopen"...

**Comment by kencu** [comment]
here's a current patch for icu, at least
```
From 7d0811ba8c7ca7403cf2ee584ef9430877184b68 Mon Sep 17 00:00:00 2001
From: ken-cunningham-webuse <ken.cunningham.webuse@gmail.com>
Date: Sat, 20 Oct 2018 14:35:08 -0700
Subject: [PATCH] icu: update to 63.1

---
 devel/icu/Portfile | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/devel/icu/Portfile b/devel/icu/Portfile
index 308b56479f4..b05ccc80fca 100644
--- a/devel/icu/Portfile
+++ b/devel/icu/Portfile
@@ -8,8 +8,8 @@ set my_name     icu4c
 # whenever the library version number changes. Thanks.
 # To find dependencies, use:
 # port file all | sort -u | xargs grep -El ':icu( |$)' | rev | cut -d / -f 2 | rev | sort -u
-version                     58.2
-# Don't update to 59 or later because they require C++11
+version                     63.1
+
 categories      devel textproc
 platforms       darwin freebsd
 maintainers                 {ryandesign @ryandesign} openmaintainer
@@ -29,8 +29,9 @@ master_sites    http://download.icu-project.org/files/${my_name}/${version}/
 distname        ${my_name}-[join [split ${version} .] _]
 extract.suffix              .tgz
 distfiles                   ${distname}-src${extract.suffix}
-checksums                   rmd160  df06e7b18a87e383d3762564f2e9a59fd75865f9 \
-                            sha256  2b0a4410153a9b20de0e20c7d8b66049a72aef244b53683d0d7521371683da0c
+checksums                   rmd160  5c895a6e2b135978df59e135ed772747aec0065f \
+                            sha256  05c490b69454fce5860b7e8e2821231674af0a11d7ef2febea9a32512998cb9d \
+                            size    23746939
 
 worksrcdir      ${name}/source
 set docdir      ${prefix}/share/doc/${name}
@@ -49,8 +50,9 @@ subport ${name}-docs {
     
     use_zip             yes
     distfiles               ${distname}-docs${extract.suffix}
-    checksums               rmd160  64c9385c43f994d283a38980c77638dc061bfe84 \
-                            sha256  dc17983debc1f23f1bd1ac1ba053f2df5b07858f5af9bd48eeb3b8044c8d3146
+    checksums               rmd160  4b967a4da3a91b44c3b1a208a88a6589f3009f67 \
+                            sha256  babb80ebad077c9e6105039657a7396812abe6818637637dab8ad4335562b763 \
+                            size    7444605
 
     extract.dir         ${worksrcpath}/doc/html
 
@@ -103,8 +105,6 @@ if { ${subport} ne "${name}-docs" } {
     }
     
     patchfiles-append   patch-config-mh-darwin.diff
-    patchfiles-append   CVE-2017-7867-CVE-2017-7868.patch
-    patchfiles-append   CVE-2017-14952.patch
     
     # ICU has three mechanisms to aid other projects in building properly
     #     1) .pc files for pkg-config
```

**Comment by michaelld** [comment]
I like the idea that this update would simplify the Portfile via fewer patches & no changed patches. Looks like there are some 90 ports that depend on icu ... any idea how many build and/or execute successfully with this update in place?

**Comment by kencu** [comment]
I think the issue was with icu-lx according to the above comments. I just tried to build it and it failed as well:
```
/opt/local/bin/clang++-mp-3.9 -dynamiclib -dynamic -O2 -pipe -Os -stdlib=libc++ -arch x86_64 -W -Wall -pedantic -Wpointer-arith -Wwrite-strings -Wno-long-long -std=c++11    -fvisibility=hidden -arch x86_64    -Wl,-compatibility_version -Wl,63 -Wl,-current_version -Wl,63.1 -install_name /opt/local/lib/libiculx.63.dylib -o ../lib/libiculx.63.1.dylib ParagraphLayout.o RunArrays.o LXUtilities.o playout.o plruns.o -L../lib -licuuc -L../stubdata -licudata -I/opt/local/include/icu-le-hb -I/opt/local/include/harfbuzz -I/opt/local/include/glib-2.0 -I/opt/local/lib/glib-2.0/include -L../lib -L/opt/local/lib -licu-le-hb -lpthread -lm 
Undefined symbols for architecture x86_64:
  "icu_63::LayoutEngine::layoutChars(char16_t const*, int, int, int, signed char, float, float, LEErrorCode&)", referenced from:
      icu_63::ParagraphLayout::ParagraphLayout(char16_t const*, int, icu_63::FontRuns const*, icu_63::ValueRuns const*, icu_63::ValueRuns const*, icu_63::LocaleRuns const*, unsigned char, signed char, LEErrorCode&) in ParagraphLayout.o
  "icu_63::LayoutEngine::layoutEngineFactory(icu_63::LEFontInstance const*, int, int, LEErrorCode&)", referenced from:
      icu_63::ParagraphLayout::ParagraphLayout(char16_t const*, int, icu_63::FontRuns const*, icu_63::ValueRuns const*, icu_63::ValueRuns const*, icu_63::LocaleRuns const*, unsigned char, signed char, LEErrorCode&) in ParagraphLayout.o
  "icu_63::LayoutEngine::getCharIndices(int*, int, LEErrorCode&) const", referenced from:
      icu_63::ParagraphLayout::ParagraphLayout(char16_t const*, int, icu_63::FontRuns const*, icu_63::ValueRuns const*, icu_63::ValueRuns const*, icu_63::LocaleRuns const*, unsigned char, signed char, LEErrorCode&) in ParagraphLayout.o
  "icu_63::LayoutEngine::getGlyphPositions(float*, LEErrorCode&) const", referenced from:
      icu_63::ParagraphLayout::ParagraphLayout(char16_t const*, int, icu_63::FontRuns const*, icu_63::ValueRuns const*, icu_63::ValueRuns const*, icu_63::LocaleRuns const*, unsigned char, signed char, LEErrorCode&) in ParagraphLayout.o
  "icu_63::LayoutEngine::getGlyphs(unsigned int*, LEErrorCode&) const", referenced from:
      icu_63::ParagraphLayout::ParagraphLayout(char16_t const*, int, icu_63::FontRuns const*, icu_63::ValueRuns const*, icu_63::ValueRuns const*, icu_63::LocaleRuns const*, unsigned char, signed char, LEErrorCode&) in ParagraphLayout.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
gnumake[1]: *** [../lib/libiculx.63.1.dylib] Error 1
```

I am not sure how many ports use `icu-lx`, but apparently it is on the way out and ports are meant to change to a `harfbuzz` implementation instead.

**Comment by kencu** [comment]
OK - forcing a rebuild of `icu-le-hb` against the new `icu`, and then building `icu-lx` worked fine:
```
$ port -v installed icu-lx
The following ports are currently installed:
  icu-lx @58.2_0 platform='darwin 10' archs='x86_64' date='2017-06-17T08:12:45-0700'
  icu-lx @63.1_0 (active) platform='darwin 10' archs='x86_64' date='2018-10-20T16:11:39-0700'
```

**Comment by kencu** [comment]
These big ones all rebuilt against the upgraded `icu`:
```
 boost @1.66.0+no_single+no_static+python27
 harfbuzz-icu @1.9.0
 qt5-qtbase @5.11.2+openssl
 openldap @2.4.46
 texlive-bin @2018.47642+x11
```

**Comment by kencu** [comment]
perhaps we might get this going again. I'll reopen.

**Comment by kencu** [comment]
I rebased this PR on the current master, worked out the conflicts, and updated icu to 63.1 instead of 61.1 -- but the commit history wound up in a detached-head state and I was unable to sort out how to reattach it to your PR. I put this work into a new PR, but if you know how to attach it here and continue this PR please feel free to do so.

**Comment by lbschenkel** [comment]
Hi, I'm busy right now and won't be able to help directly but I don't mind at the slightest for you to continue on a different PR. It'll be great if you manage to get ICU updated.

**Comment by michaelld** [comment]
Since you referenced this PR in the next one, my vote is to close this one & make the new one the active one.

**Comment by lbschenkel** [comment]
I thought this was closed. I'm closing it.

**Comment by jmroot** [comment]
I guess this could be revisited now that the icu-bootstrap subport exists and is used to get a C++11 compiler on 10.6. The main icu port could be updated to the latest version, while leaving icu-bootstrap at the last pre-C++11 version.

**Comment by lbschenkel** [comment]
I think it would be great to get ICU updated, however I no longer feel welcome to work on WIP pull requests for MacPorts. If the work on this PR is of any value in a new attempt, please feel encouraged to take it and make use of it.

**Pull request closed without merging**